src/librbd/io/AioCompletion.cc

```diff
@@ -250,9 +252,16 @@ void AioCompletion::complete_external_callback() {
 
     AioCompletion* aio_comp;
     while (ictx->external_callback_completions.pop(aio_comp)) {
+      // ensure we have at least one valid AioCompletion not finalized while
+      // looping to prevent ImageCtx from being closed
```

Comment:
I am still not quite sure it is not racy. Consider a case when we did `ictx->external_callback_completions.push(this)` and then another thread processed all completions before we tried `ictx->external_callback_in_progress.exchange(true)`? I.e. theoretically `ictx` might be invalid here.

Reply:
I moved the `async_op.finish_op()` to a new `destroy` method which will only be invoked when the last reference is dropped. Alternatively I can just always enqueue the final completion in the `op_work_queue` and avoid the concern about the added latency from the extra thread bounce.

Reply:
Cancel that -- I am just going to mimic the master-branch technique to avoid octopus being wildly divergent. It will just cost and extra heap allocation to complete.